### PR TITLE
Peter/basic api functions

### PR DESCRIPTION
**Pull request by Ripiter**

### CODE REVIEW

File: Source/IdentityApi/IdentityApi/Controllers/UserController.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        private readonly ITokenManager _tokenManager;
+        public UserController(IUserManager userManager, ITokenManager tokenManager)
+        {
+            _userManager = userManager;
+            _tokenManager = tokenManager;
+        }
+
+        [HttpPost]
+        [Route("Login")]
+        public async Task<ActionResult<UserToken>> Login(UserLogin userLogin)
```
Comment (blackout1471) [inline Source/IdentityApi/IdentityApi/Controllers/UserController.cs]: Comment

File: Source/IdentityApi/IdentityApi/Controllers/UserCreateController.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        private readonly ITokenManager _tokenManager;
+        public UserCreateController(IUserManager userManager, ITokenManager tokenManager)
+        {
+            _userManager = userManager;
+            _tokenManager = tokenManager;
+        }
+
+        [HttpPost]
+        [Route("Register")]
+        public async Task<ActionResult<UserToken>> Register(UserCreate userCreate)
```
Comment (blackout1471) [inline Source/IdentityApi/IdentityApi/Controllers/UserCreateController.cs]: Comment

File: Source/IdentityApi/IdentityApi/Helpers/Security.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+namespace IdentityApi.Helpers
+{
+    public static class Security
+    {
+        /// <summary>
+        /// Uses Sha512 to compute the hash of the password
+        /// </summary>
+        /// <param name="password"></param>
+        /// <returns></returns>
+        private static string GetEncryptedPassword(string password)
```
Comment (blackout1471) [inline Source/IdentityApi/IdentityApi/Helpers/Security.cs:12]: Fake it till you make it

File: Source/IdentityApi/IdentityApi/Managers/TokenManager.cs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        private readonly IConfiguration _configuration;
+        private readonly string _secret;
+        public TokenManager(IConfiguration configuration, ILogger<TokenManager> logger)
+        {
+            _configuration = configuration;
+            _logger = logger;
+            _secret = _configuration["JWT:Key"];
+        }
+
+        public UserToken GetUserTokenFromToken(string token)
```
Comment (blackout1471) [inline Source/IdentityApi/IdentityApi/Managers/TokenManager.cs:23]: <inheritDoc />

Comment (blackout1471) [reply Source/IdentityApi/IdentityApi/Managers/TokenManager.cs:23]: inheritdoc

File: Source/IdentityApi/IdentityApi/Managers/UserManager.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+{
+    public class UserManager : IUserManager
+    {
+        private readonly IUserProvider _userProvider;
+        public UserManager(IUserProvider userProvider)
+        {
+            _userProvider = userProvider;
+        }
+
+        public async Task<User> CreateUser(UserCreate userCreate)
```
Comment (blackout1471) [inline Source/IdentityApi/IdentityApi/Managers/UserManager.cs]: inheritdoc

File: Source/IdentityApi/IdentityApi/Managers/UserManager.cs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                Email = createdUser.Email,
+                FirstName = createdUser.FirstName,  
+                LastName = createdUser.LastName,
+                PhoneNumber = createdUser.PhoneNumber
+            };
+        }
+
+        public async Task<User> Login(UserLogin userLogin)
+        {
+            // checjs uf yser exuts
```
Comment (blackout1471) [inline Source/IdentityApi/IdentityApi/Managers/UserManager.cs]: yes

File: Source/IdentityApi/IdentityApi/Models/UserLocation.cs [hunk 7]
```diff
@@ -0,0 +1,6 @@
+﻿namespace IdentityApi.Models
+{
+    public class UserLocation
```
Comment (blackout1471) [inline Source/IdentityApi/IdentityApi/Models/UserLocation.cs:3]: D afuckieralksd

File: Source/IdentityApi/IdentityApi/Providers/UserProvider.cs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    public class UserProvider : IUserProvider
+    {
+        private readonly IConfiguration _configuration;
+
+        public UserProvider(IConfiguration configuration)
+        {
+            _configuration = configuration;
+        }
+
+        public async Task<DbUser> CreateUser(DbUser userCreate)
```
Comment (blackout1471) [inline Source/IdentityApi/IdentityApi/Providers/UserProvider.cs]: inheritdoc

File: Source/IdentityApi/IdentityApiTest/Managers/UserManagerTest.cs [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                Password = "abc",
+                FirstName = "jon",
+                LastName = "stevensen",
+                PhoneNumber = "13246578"
+            };
+        }
+
+        [Fact]
+        public void User_Should_Login()
+        {
```
Comment (blackout1471) [inline Source/IdentityApi/IdentityApiTest/Managers/UserManagerTest.cs]: // arrange, act, assert

File: Source/IdentityApi/IdentityApiTest/Managers/UserManagerTest.cs [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+            var userManager = new UserManager(userProvider);
+            var actual =  userManager.Login(userLogin).Result;
+
+            Assert.Equal(expected.Email, actual.Email);
+            Assert.Equal(expected.ID, actual.ID);
+        }
+
+        [Fact]
+        public void User_Should_ThrowOnLogin()
```
Comment (blackout1471) [inline Source/IdentityApi/IdentityApiTest/Managers/UserManagerTest.cs]: Refactor names WhatToExpect_When_ActionOccurs ex: ThrowException_When_UserLogin

**Review by blackout1471 — COMMENTED** [review]
All async methods should also be called async at the end

File: Docs/ClassDiagrams/ApiClassDiagram.plantuml [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+TokenManager .-|> ITokenManager
+ITokenManager *-- UserToken
+
+
+'Providers'
+
+interface IUserProvider{
+    +CreateUserAsync(userCreate : DbUser) : Task<DbUser>
+    +GetUserByEmailAsync(email : string) : Task<DbUser>
+    +GetuserByIdAsync(id : int) : Task<DbUser>
```
Comment (Geefy) [inline Docs/ClassDiagrams/ApiClassDiagram.plantuml]: GetuserByIdAsync -> GetUserByIdAsync

File: Docs/SDDiagrams/CreateUserSD.plantuml [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+alt User found
+
+UserManager->UserController: User already exists
+UserController->User: User already exists
+
+end
+
+
+UserManager->Security: Get salt
+Security->User: Return salt
```
Comment (Geefy) [inline Docs/SDDiagrams/CreateUserSD.plantuml]: Should point back to UserManager and not User

**Review by Geefy — CHANGES_REQUESTED** [review]

**Review by blackout1471 — APPROVED** [review]

**Review by Geefy — APPROVED** [review]

**Pull request merged**